### PR TITLE
Add AppArmor confinement for Qemu, Qemu-img and DNSMasq

### PR DESCRIPTION
**Pull request by gerboland**
- ProcessFactory will return AppArmor-wrapped Processes if using qemu backend or DISABLE_APPARMOR env var set.
 - On creation these Processes install the AppArmor profile defined in the ProcessSpec, and executes the binary within this confinement. On Process deletion, the AppArmor profile is uninstalled.
 - Add AppArmor profiles for Qemu, Qemu-img and DNSMasq

For testing, I needed to add mpt::ResetProcessFactory to ensure the ProcessFactory is reset before & after the test suite runs.

---
Tip for reviewing:
- Use `sudo aa-status` to check the list of AppArmor profiles loaded. The multipass profiles will be in enforce mode (except multipassd, as it's classicly confined still). PIDs being confined are listed after the profile name.

### CODE REVIEW

**Comment by Saviq** [comment]
Available in the `edge/strict` channel now: `edge/strict  0.9.0-dev.84+g5702c72  1082`

**Comment by gerboland** [comment]
CI unhappy I believe due to #995

**Comment by ricab** [comment]
> CI unhappy I believe due to #995

#995 is this pull :)

**Comment by gerboland** [comment]
D'oh, darn autocomplete. I meant #998  Thanks @ricab

**Comment by gerboland** [comment]
Good to review, only Mac failing in CI due to conflict

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -21,6 +21,14 @@ architectures:
 - build-on: arm64
   run-on: arm64
 
+layout:
```
[the remark this answers is not in the document]
Comment (gerboland) [reply snap/snapcraft.yaml]: I can't use a `*` in the definition, but turns out `$SNAPCRAFT_ARCH_TRIPLET` can be used here.

File: src/platform/backends/qemu/dnsmasq_process_spec.cpp [hunk 2]
```diff
@@ -17,9 +17,11 @@
 
 #include "dnsmasq_process_spec.h"
 
+#include <multipass/snap_utils.h>
 #include <multipass/format.h>
 
 namespace mp = multipass;
+namespace mu = multipass::utils;
```
[the remark this answers is not in the document]
Comment (gerboland) [reply src/platform/backends/qemu/dnsmasq_process_spec.cpp]: Can I leave that to a later MP, as this precedent is already in master? I don't want to make this PR any bigger

File: src/platform/backends/shared/linux/linux_process.cpp [hunk 3]
```diff
@@ -17,13 +17,23 @@
 
 #include "linux_process.h"
 #include "process_spec.h"
+#include <QProcess>
```
[the remark this answers is not in the document]
Comment (gerboland) [reply src/platform/backends/shared/linux/linux_process.cpp]: Nice catch

File: src/platform/backends/qemu/qemu_virtual_machine.cpp [hunk 4]
```diff
@@ -272,9 +272,16 @@ mp::QemuVirtualMachine::QemuVirtualMachine(const VirtualMachineDescription& desc
         }
         if (process_state.error)
         {
-            mpl::log(
-                mpl::Level::error, vm_name,
-                fmt::format("process returned error {}: {}", process_state.error->state, process_state.error->message));
+            if (process_state.error->state == QProcess::Crashed &&
+                (state == State::suspending || state == State::suspended))
+            {
+                // when suspending, we ask Qemu to savevm. Once it confirms that's done, we kill it. Catch the "crash"
```
[the remark this answers is not in the document]
Comment (gerboland) [reply src/platform/backends/qemu/qemu_virtual_machine.cpp]: Sure. The crash was throwing me off until I figured out why.

File: src/platform/backends/shared/linux/process_factory.cpp [hunk 5]
```diff
@@ -47,21 +42,85 @@ class SimpleProcessSpec : public mp::ProcessSpec
...
+};
+
+mp::optional<mp::AppArmor> create_apparmor()
+{
+    if (qEnvironmentVariableIsSet("DISABLE_APPARMOR"))
+    {
+        mpl::log(mpl::Level::warning, "apparmor", "AppArmor disabled by environment variable");
+        return mp::nullopt;
+    }
+    else if (mp::utils::get_driver_str() == "libvirt")
```
[the remark this answers is not in the document]
Comment (gerboland) [reply src/platform/backends/shared/linux/process_factory.cpp]: The libvirt daemon applies its own apparmor security profiles. So we don't need to do anything.

[the remark this answers is not in the document]
Comment (gerboland) [reply src/platform/backends/shared/linux/process_factory.cpp]: Yep, that won't be hard to do. In theory just removing this else-if statement will do - but I wanted to confine changes to one backend at a time - eases testing load.

**Comment by ricab** [comment]
@townsend2010 I'd have to allocate a whole chunk of time to parse this, so I'd leave this one up to you if that is ok...

**Comment by townsend2010** [comment]
@ricab, ack, will just approve then.

**Comment by townsend2010** [comment]
Let's do it!

bors r+

**Pull request merged**